### ISSUE TITLE
Java incorrect result: array<datatype> issues with equality checking seq 

### ISSUE THREAD

**Issue by Dilan-s** [dafny-lang/dafny#4032]
### Dafny version

4.1.0

### Code to produce this issue

```dafny
datatype DT_1 = DT_1_1 | DT_1_2

method Main() returns ()
{
	var v_DT_1_2_8: DT_1 := DT_1_2;
	var v_DT_1_2_9: DT_1 := DT_1_2;
	var v_DT_1_2_10: DT_1 := DT_1_2;
	var v_array_7: array<DT_1> := new DT_1[3] [v_DT_1_2_8, v_DT_1_2_9, v_DT_1_2_10];
	var v_DT_1_2_11: DT_1 := DT_1_2;
	var v_DT_1_2_12: DT_1 := DT_1_2;
	var v_DT_1_2_13: DT_1 := DT_1_2;
	var v_array_8: array<DT_1> := new DT_1[3] [v_DT_1_2_11, v_DT_1_2_12, v_DT_1_2_13];
	var v_seq_3: seq<array<DT_1>> := ([v_array_7, v_array_8]);

	print "v_seq_3", " ", (v_seq_3 == [v_array_7, v_array_8]), "\n";
	v_array_8 := v_array_7;
	print "v_seq_3", " ", (v_seq_3 == [v_array_7, v_array_8]), "\n"; // CURRENT ISSUE

	print "v_seq_3[0][0]", " ", v_seq_3[0][0], " ", "v_array_7[0]", " ", v_array_7[0], "\n";
	v_seq_3[0][0] := DT_1_1;
	print "v_seq_3[0][0]", " ", v_seq_3[0][0], " ", "v_array_7[0]", " ", v_array_7[0], "\n"; // POTENTIAL ISSUE
}
```

### Command to run and resulting output

```code
$ Dafny /compile:4 /compileVerbose:0 /compileTarget:cs test.dfy

Dafny program verifier finished with 1 verified, 0 errors
v_seq_3 true
v_seq_3 false
v_seq_3[0][0] DT_1.DT_1_2 v_array_7[0] DT_1.DT_1_2
v_seq_3[0][0] DT_1.DT_1_1 v_array_7[0] DT_1.DT_1_1


$ Dafny /compile:4 /compileVerbose:0 /compileTarget:java test.dfy

Dafny program verifier finished with 1 verified, 0 errors
v_seq_3 true
v_seq_3 true
v_seq_3[0][0] DT_1.DT_1_2 v_array_7[0] DT_1.DT_1_2
v_seq_3[0][0] DT_1.DT_1_1 v_array_7[0] DT_1.DT_1_1
```

### What happened?

Running the program above in Java produces the result with the final comparison giving the value true, however in all other backends (cs, js, py and go) the final comparison produces the value false.

The expected result depends on the intention of Dafny when adding a value into a seq. If the value is shallow copied, then I would expect the reference to v_array_8 within v_seq_3 to be adjusted when the assignment is performed, and hence the result to be true. If the value is deep copied, then I would expect the reference to v_array_8 within v_seq_3 not to be changed, and hence false produced. 

The final print in the code above is used to verify this behaviour remains consistent based upon the choice take.

As a note with this, changing to array\<int\> does not produce an error.

### What type of operating system are you experiencing the problem on?

Linux

**Comment by Dilan-s** [comment]
From investigation, this seems to be an issue with the cast being done in TypeDescriptor.java
```java
        @Override
        public boolean arrayDeepEquals(Object array1, Object array2) {
            @SuppressWarnings("unchecked")
            T[] castArray1 = (T[]) array1;
            @SuppressWarnings("unchecked")
            T[] castArray2 = (T[]) array2;
            return Arrays.deepEquals(castArray1, castArray2);
        }
```

The use of Arrays.deepEquals checks the elements of the array (which happen to be the same in v_array_7 and v_array_8) and hence return true, while it should be checking if the objects in the array are also the same memory references.

**Comment by keyboardDrummer** [comment]
Closing because this issue is resolved in Dafny 4.6
